### ISSUE TITLE
set x-frame-options

### ISSUE THREAD

**Issue by diracdeltas** [brave/brave-website#13]
we don't need to be iframed, so we should set https://developer.mozilla.org/en-US/docs/Web/HTTP/X-Frame-Options to DENY

**Comment by retzion** [comment]
Looks like this will do it in hapi { security: { xframe: false } }

**Comment by retzion** [comment]
A little more on that..

            var server = new Hapi.Server({ security: { xframe: false } });
            server.route({ method: 'GET', path: '/', handler: handler });

            server.inject({ url: '/' }, function (res) {

                expect(res.result).to.exist;
                expect(res.result).to.equal('Test');
                expect(res.headers['x-frame-options']).to.not.exist;
                expect(res.headers['strict-transport-security']).to.equal('max-age=15768000');
                expect(res.headers['x-xss-protection']).to.equal('1; mode=block');
                expect(res.headers['x-download-options']).to.equal('noopen');
                expect(res.headers['x-content-type-options']).to.equal('nosniff');
                done();
            });

I can give it a go if you want. Just assign the issue to me.

**Comment by retzion** [comment]
I think this will do the trick in hapi.

            var server = new Hapi.Server({ security: { xframe: true } });
            server.route({ method: 'GET', path: '/', handler: handler });

            server.inject({ url: '/' }, function (res) {
                expect(res.result).to.exist;
                expect(res.headers['x-frame-options']).to.equal('DENY');
                done();
            });

Assign me the task if you want me to try.

**Comment by diracdeltas** [comment]
yep, thanks. @retzion you should feel free to take any task that isn't assigned to anyone. :)

**Comment by diracdeltas** [comment]
Please set HSTS (HTTP Strict Transport Security headers) to` Strict-Transport-Security: max-age=31536000; includeSubDomains` while you're at it

**Comment by diracdeltas** [comment]
@bridiver any preference if we set these in Fastly vs in hapi?

**Comment by bridiver** [comment]
Generally easier to set headers at the origin (hapi), but will that cause problems for testing on localhost or otherwise connecting directly to heroku where there will be a domain mismatch?

**Comment by diracdeltas** [comment]
it might but i think that is outweighed by the benefit of having the headers be CDN-independent, so let's go with hapi as @retzion originally planned. worst-case devs can edit /etc/hosts as needed.

**Comment by bridiver** [comment]
We could also set them by node_env. We don't currently have that set in Heroku, but it would be easy to add.

**Comment by retzion** [comment]
So the first code sample I posted here produced nothing but errors.
I came up with the following changes and they are currently staged.

```
  server.route({
    method: 'GET',
    path: '/{path*}',
    config: {
      state: {
        parse: true, 
        failAction: 'log' 
      }
      , security: {
        hsts: {
          maxAge: 31536000,
          includeSubDomains: true,
          preload: true
        }
        , xframe: true
      }
    },
    handler: {
      directory: {
        path: './public'
      }
    }
  })


  server.ext('onRequest', function (request, reply) {
      if (request.headers['x-forwarded-proto'] != 'https') {
    console.log('request.headers[x-forwarded-proto]',request.headers['x-forwarded-proto'])
    console.log('process.env.PORT',process.env.PORT)
    console.log('redirecting.. ','https://' + request.headers.host + request.url.path)
        return reply()
          .redirect('https://' + request.headers.host + request.url.path)
          .code(301);
      }
      reply.continue()      
  })
```


I have this running on a fresh Heroku app install at http://brave-website-ru.herokuapp.com
The redirect works fine via hsts. However things are not working for http://evarb.brave.com
Any ideas, here?
I would like to see some server logs from the Heroku app if you can produce some. Thanks

**Comment by retzion** [comment]
@diracdeltas do we want preload:true?
preload - a boolean specifying whether to add the 'preload' flag (used to submit domains inclusion in Chrome's HTTP Strict Transport Security (HSTS) preload list) to the header.

**Comment by retzion** [comment]
Hmm, seems to be working now. http://evarb.brave.com should bounce to https

**Comment by marumari** [comment]
@bridiver: Don't know whether it answers your question or not, but it's generally hard to screw things up with regards to HSTS and localhost.  Browsers will ignore the flag if it is sent over an insecure connection, so setting HSTS on stuff delivered over http on localhost won't do anything at all.

**Comment by diracdeltas** [comment]
@retzion, yeah please set preload on hsts

**Comment by retzion** [comment]
I committed with Preload: true
I think we can close this issue now.

**Comment by retzion** [comment]
I'm still trying to get a self-signed cert working on localhost. Can anyone confirm if things work on local?

**Comment by retzion** [comment]
Been using this..

    // temporary workaround to allow localhost with HSTS
    var connections = [
          // non-local host
          {
              port: process.env.PORT || 3000,
          },
          // localhost
          {
              host: 'localhost',
              port: 3001,
              tls: {
                key: fs.readFileSync('./server.key'),     // Path to key
                cert: fs.readFileSync('./server.crt'),      // Path to self-signed Certificate

                // This is necessary only if using the client certificate authentication.
                requestCert: true,

                // This is necessary only if the client uses the self-signed certificate.
                ca: []
            }
          }
      ]

    // IMPORTANT : If you toggle Connections, toggle back to [0] if not localhost
    server.connection(connections[1])

...

    server.ext('onRequest', function (request, reply) {
      let isLocalhost = ((request.headers.host.substring(0,request.headers.host.indexOf(':'))) == 'localhost')
      console.log('isLocalhost',isLocalhost)
        if (request.headers['x-forwarded-proto'] !== 'https' && !isLocalhost) {
          return reply()
            .redirect('https://' + request.headers.host + request.url.path)
            .code(301);
        }
        reply.continue()      
    })
